### ISSUE TITLE
Enhanced documentation or support for deprecated/obsoleted packages

### ISSUE THREAD

**Issue by monty241** [semver/semver#983]
Based upon [V2.0 of Semantic Version](https://semver.org/spec/v2.0.0.html), there is no specific way to signal that a release numbering following Semantic Version is "deprecated", with "deprecated" meaning that it is end-of-life and users are urged to upgrade. 

A deprecated/obsoleted product is still often fine for production use.

An alternative would be to use for instance "-depr" as a "prerelease", but that signals also a pre-release to for instance the related tools of NuGet. Build metadata would also be an option, which are new with 2.0.0, but less well supported.

For now, we are sticking to including "-depr" in the file name of NuGet packages, but having a version number without "-depr" in the NuGet package metadata.

It would be great when the standard could be extended to explicitly state how to signal that the whole library, package, executable or package is in an deprecated state.

Related:

- https://github.com/semver/semver/issues/325

**Comment by ljharb** [comment]
A semver version can’t ever change; it’s immutable. Deprecation status changes. As such, it’s not appropriate to couple the two - how to handle deprecation should be determined by an individual package manager.

**Comment by monty241** [comment]
Yes, that is correct. Our releases, say 22.0.1 - 22.0.705, were production, but everything released AFTER the next major version hits the market is signaled as deprecated, so 22.0.706 - 22.0.813, are releases that are already deprecated on initial release, but serve to give users time to migrate to the next major release.

Similar to what Microsoft does with Extended Support and other End of Life Support packages.

**Comment by ljharb** [comment]
Interesting, that’s a very strange approach to me. Either way I’d say the deprecation status should always be independent of the version number.

**Comment by ljharb** [comment]
It seems like prereleases of the *next* major would be a clearer way to signal that, though?

**Comment by monty241** [comment]
Yes, I can imagine that it is atypical, but I see it often in business software (so large to very large software packages). Beyond that there are releases, there is even a ecosystem around dead software like Spinnaker Support that keep them alive say for five years after terminal releases.

Support is typically given only on some versions, typically in "Production" state, but when using deprecated versions it is a strong signal that a product has limited or no remaining regular supported lifetime.

A sample of this is the ESU ("Extended Security Update") program of Microsoft.

With often multiple major releases in the field, for instance 3 overlapping in varying grades of support, I don't know how to signal with the prerelease of the next major what major will be phased out.

**Comment by jwdonahue** [comment]
With the exception of prerelease, there's nothing in SemVer to express policies.  That sort of thing should be handled at the package feed level, not in the version number.  There's potentially infinite policies that might apply to a particular publisher's offerings, so no good way to standardize on that in a short, universally understood string. 

I've seen some schemes in the past that encoded non-version information, along with the version string, mostly for ROMs or circuit boards, but they all have proprietary meanings.

**Comment by jwdonahue** [comment]
I would add that it's not always possible to know in advance, that a particular version is going to be the last of that series.

**Comment by monty241** [comment]
Yes, we are now using something proprietary. Unsure whether other publishers of large packages have the same itch. 

Regarding knowing in advance: it is not possible nor desired to know in advance unless there is some kind of formal deprecation list or certification matrix employed by companies like Oracle, IBM or Microsoft. In our case, the deprecated state finds its way back in the legal agreements and product lifetime agreements.

**Comment by jwdonahue** [comment]
In your case, if you package tools and feeds preserve the SemVer build meta tag, you could just add a +EOL or +FinalFix tag to whatever release you expect not to follow up with further bug fixes.  That's the kind of thing you could easily fold into your contracts.

**Comment by jwdonahue** [comment]
Based on [this MS-NuGet blog](https://devblogs.microsoft.com/nuget/supporting-semver-2-0-0/) and some discussions we had with that team at MS a long while back (I was on the Windows build team then), I am pretty sure that the entire MS supported tool chain and feeds for nupkg's, has full support for build meta tags in the version string.
